005-1-locations-on-stack.md

```diff
@@ -0,0 +1,590 @@
+> space and pushed onto the stack.
+>
+> 3. `DW_OP_form_tls_address`... [moved from Section 2.5.1.3]
+> 
+> 4. `DW_OP_call_frame_cfa`... [moved from Section 2.5.1.3]
+
+
+### Section 2.5.2.2 Register Locations
+
+Move the contents of Section 2.6.1.1.3 here, replacing the term
+"location description" with "location" throughout.
+
+[page 39] Remove the non-normative text:
+
+> A register location description must stand alone as the entire
+> description of an object or a piece of an object.
+
+
+### Section 2.5.2.3 Implicit Locations
+
+Move the contents of Section 2.6.1.1.4 here, replacing the term
+"location description" with "location" throughout.
+
+
+### Section 2.5.2.4 Undefined Locations
+
+Insert the following (adapted from Section 2.6.1.1.1):
+
+> An undefined location represents a piece or all of an object that is
+> present in the source but not in the object code (perhaps due to
+> optimization). The `DW_OP_undefined` operation pushes an undefined
+> location onto the stack. A DWARF expression containing no operations or
+> that leaves no elements on the stack also produces an undefined
+> location.
+
+
+### Section 2.5.2.5 Composite Locations
+
+Insert the following (adapted from Section 2.6.1.2, and with the new
+`DW_OP_composite` operator):
+
+> The above kinds of locations are considered "simple" locations.
+> 
+> A composite location describes an object or value which may be
+> contained in part of a register or stored in more than one location.
+> Each piece is described by a composition operation. There may be one
+> or more composition operations in a single composite location
+> description.
+>
+> A series of piece operations (`DW_OP_piece` or `DW_OP_bit_piece`)
+> describes the parts of a value in memory address order. Each piece
+> operation pops a location `A` from the stack and updates the composite
+> location `B` in the preceding element of the stack by appending the
+> new piece described by `A`.
+>
+> A composite location may be formed from several simple locations by
+> the composition operations described in this section. Each simple
+> location describes the location of one piece of the object; each
+> composition operation describes which part of the object is located
+> there.
+> 
+> 1. `DW_OP_composite`
+>
+>     The `DW_OP_composite` operator has no operands. It pushes a new, empty,
+>     composite location onto the stack.
+>
+>     _This operator is provided so that a new series of piece operations
+>     can be started to form a composite location when the state of the
+>     stack is unknown (e.g., following a `DW_OP_call*` operation)._
+> 
+> 2. `DW_OP_piece`
+>
+>     The `DW_OP_piece` operation takes a single operand, which is an unsigned
+>     LEB128 number. The number describes the size `S`, in bytes, of the piece
+>     of the object referenced by the location `A` on the top of the stack. If
+>     the piece is located in a register, but does not occupy the entire
```

Comment:
> If the piece is located in a register, but does not occupy the entire register, the placement of the piece within that register is defined by the ABI.

Yeah, so in our ideal world where we treat all storages uniformly,  just as sequences of bits, we wouldn't have that.  Perhaps a point to discuss.

In DWARF 5, register location descriptions just specified the whole register, so I guess a consumer could interpret: "when you say reg 2, I magically know that you mean those specific bytes of reg 2".

But now that register locations have an offset component, I'm not sure how a consumer is supposed to interpret these:

    DW_OP_composite
    DW_OP_reg2
    DW_OP_piece 2

vs

    DW_OP_composite
    DW_OP_reg2
    DW_OP_offset 1
    DW_OP_piece 2

In the first one, the register location has an offset of 0.  So I guess that consumers in that specific case (offset == 0) could somehow *hand waving* decide to apply the same "placement defined by ABI" rules they applied in DWARF 5.

But then, what about the second example?  The producer specified an explicit offset, so it would be strange to just ignore it and pick some other bytes.

And then, what if the producer really wants to point at those bytes at offset 0 in register 2?  If some magic "placement defined by ABI" rules kick in, it's just not possible.

I recall this presentation from Andreas Arnez, if I recall correctly it's a concrete example of this problem.

https://youtu.be/iQAd5Atlz1s?list=PL_GiHdX17Wtx2Bu1O_bREetZZv4moIaRi&t=2137

And also these threads:

`[Dwarf-Discuss] DWARF piece questions`
https://www.mail-archive.com/dwarf-discuss@lists.dwarfstd.org/msg00344.html

`[gdb] [RFC] DW_OP_piece vs. DW_OP_bit_piece on a Register`
https://inbox.sourceware.org/gdb/m3vb6wm86q.fsf@oc1027705133.ibm.com/

If time permits, I'd like to go through them again to understand what impact our proposal would have for the problems he presents.

Reply:
Yes, this is the point I was trying to get at in my "registers really are different" arguments and my Feb. 27 email about "offsets".